src/aws/flb_aws_util.c

```diff
@@ -507,3 +512,190 @@ int flb_read_file(const char *path, char **out_buf, size_t *out_size)
+    }
+
+    tmp_tag = flb_sds_create_len(tag, strlen(tag));
+    if(!tmp_tag){
+        goto error;
+    }
+
+    s3_key = flb_sds_create_len(format, strlen(format));
+    if (!s3_key) {
+        goto error;
+    }
+
+    /* Check if delimiter(s) specifed exists in the tag. */
+    for (i = 0; i < strlen(tag_delimiter); i++){
+        if (strchr(tag, tag_delimiter[i])){
+            ret = 1;
+            break;
+        } 
+    }
+
+    tmp = flb_sds_create_len(TAG_PART_DESCRIPTOR, 5);
+    if (!tmp) {
+        goto error;
+    }
+    if (strstr(s3_key, tmp)){
+        if(ret == 0){
+            flb_warn("[s3_key] Invalid Tag delimiter: does not exist in tag. "
+                    "tag=%s, format=%s", tag, format);
+        }  
+    }
+
+    flb_sds_destroy(tmp);
+
+    /* Split the string on the delimiters */
+    tag_token = strtok(tmp_tag, tag_delimiter);
+    
+    /* Find all occurences of $TAG[*] and  
+     * replaces it with the right token from tag.
+     */
+    while(tag_token != NULL && i < MAX_TAG_PARTS) {
+        buf = flb_sds_create_size(10);
+        if (!buf) {
+            goto error;
+        }
+        tmp = flb_sds_printf(&buf, TAG_PART_DESCRIPTOR, i);
+        if (!tmp) {
+            goto error;
+        }
+
+        tmp_key = replace_uri_tokens(s3_key, tmp, tag_token);
+        if (!tmp_key) {
+            goto error;
+        }
+
+        if(strlen(tmp_key) > S3_KEY_SIZE){
+            flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+        }
+
+        flb_sds_destroy(tmp);
+        flb_sds_destroy(s3_key);
+        s3_key = tmp_key;
+
+        tag_token = strtok(NULL, tag_delimiter);
+        i++;
+    }
+    
+    tmp = flb_sds_create_len(TAG_PART_DESCRIPTOR, 5);
+    if (!tmp) {
+        goto error;
+    }
+
+    /* A match against "$TAG[" indicates an invalid or out of bounds tag part. */
+    if (strstr(s3_key, tmp)){
+        flb_warn("[s3_key] Invalid / Out of bounds tag part: At most 10 tag parts "
+                 "($TAG[0] - $TAG[9]) can be processed. tag=%s, format=%s", tag, format);
+    }
+
+    /* Find all occurences of $TAG and replace with the entire tag. */
+    tmp_key = replace_uri_tokens(s3_key, TAG_DESCRIPTOR, tag);
+    if (!tmp_key) {
+        goto error;
+    }
+
+    if(strlen(tmp_key) > S3_KEY_SIZE){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+    }
+
+    flb_sds_destroy(s3_key);
+    s3_key = tmp_key; 
+
+    gmt = gmtime(&time);
+
+    flb_sds_destroy(tmp);
+
+    /* A string longer than S3_KEY_SIZE is created to store the formatted timestamp. */ 
+    tmp = flb_sds_create_size(S3_KEY_SIZE);
+    if (!tmp) {
+        goto error;
+    }
+
+    ret = strftime(tmp, S3_KEY_SIZE, s3_key, gmt);
+    if(ret == 0){
+        flb_warn("[s3_key] Object key length is longer than the 1024 character limit.");
+    }
+
+    flb_sds_destroy(s3_key);
+    s3_key = tmp;
+
+    flb_sds_destroy(tmp_tag);
+    return s3_key;
```

Comment:
I ran into a small issue with this code today; the flb_sds_t object returned does not have its length properly set, since strftime treats it as a generic buffer or characters.

To fix this, just set the length before returning. This will ensure all of the standard sds functions work on the returned string:
```
FLB_SDS_HEADER(s3_key)->len = strlen(s3_key);
return s3_key;
```

Reply:
I did not know about this. Thank you